python/ray/serve/backend_state.py

```diff
@@ -78,35 +113,139 @@ def get_replica_handles(
     def get_backend(self, backend_tag: BackendTag) -> Optional[BackendInfo]:
         return self.backends.get(backend_tag)
 
+    def num_pending_goals(self) -> int:
+        return len(self.pending_goals)
+
+    async def wait_for_goal(self, goal_id: GoalId) -> None:
+        start = time.time()
+        if goal_id not in self.pending_goals:
+            logger.debug(f"Goal {goal_id} not found")
+            return True
+        event = self.pending_goals[goal_id]
+        await event.wait()
+        logger.debug(
+            f"Waiting for goal {goal_id} took {time.time() - start} seconds")
+
+    def _complete_goal(self, goal_id: GoalId) -> None:
+        logger.debug(f"Completing goal {goal_id}")
+        event = self.pending_goals.pop(goal_id, None)
+        if event:
+            event.set()
+
+    def _create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
```

Comment:
I think this would actually a little bit of a leaky abstraction. The goal-related interfaces don't need to know that this happens when a goal is recreated (in fact they don't need to know that recreation happens at all). e.g., imagine that we move this into a separate `AsyncGoalManager` class (which we should) -- it should either handle checkpoint/recovery entirely on its own (so this API wouldn't be used to recreate goals) or it should just expose a generic interface that's used for checkpoint/recovery.

Reply:
Obviously not very important, that's just why I renamed it to be more generic :)

Reply:
That makes sense! I agree that the *entire* goal handling process should be as generic as possible. I think abstracting it into a different class would be super helpful and would make the code much more clear.